// @pact-foundation/pact-node/dist/standalone/setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setup = void 0;
var path = require("path");
var pact_environment_1 = require("../src/pact-environment");
var config_1 = require("./config");
var urls_1 = require("./urls");
function setup(platform, arch) {
    var entries = config_1.getBinaryEntries(platform, arch);
    return Promise.resolve(entries.map(function (entry) { return ({
        binaryDownloadPath: urls_1.join(entry.downloadLocation, entry.binary),
        checksumDownloadPath: urls_1.join(config_1.PACT_DEFAULT_LOCATION, entry.binaryChecksum),
        filepath: path.resolve(__dirname, entry.binary),
        checksumFilepath: entry.binaryChecksum === 'skip'
            ? 'skip'
            : path.resolve(__dirname, entry.binaryChecksum),
        isWindows: pact_environment_1.default.isWindows(platform),
        platform: entry.platform,
        arch: entry.arch,
        platformFolderPath: path.resolve(__dirname, entry.folderName),
    }); }));
}
exports.setup = setup;
