// @pact-foundation/pact-node/dist/standalone/connectors/tracking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.trackDownload = void 0;
var chalk_1 = require("chalk");
var config_1 = require("../config");
var versions_1 = require("../versions");
var request_1 = require("./boundaries/request");
var CIs = [
    'CI',
    'CONTINUOUS_INTEGRATION',
    'ABSTRUSE_BUILD_DIR',
    'APPVEYOR',
    'BUDDY_WORKSPACE_URL',
    'BUILDKITE',
    'CF_BUILD_URL',
    'CIRCLECI',
    'CODEBUILD_BUILD_ARN',
    'CONCOURSE_URL',
    'DRONE',
    'GITLAB_CI',
    'GO_SERVER_URL',
    'JENKINS_URL',
    'PROBO_ENVIRONMENT',
    'SEMAPHORE',
    'SHIPPABLE',
    'TDDIUM',
    'TEAMCITY_VERSION',
    'TF_BUILD',
    'TRAVIS',
    'WERCKER_ROOT',
];
var trackDownload = function (platform) {
    if (!config_1.CONFIG.doNotTrack) {
        console.log(chalk_1.default.gray('Please note: we are tracking this download anonymously to gather important usage statistics. ' +
            "To disable tracking, set 'pact_do_not_track: true' in your package.json 'config' section."));
        var isCI = CIs.some(function (key) { return process.env[key] !== undefined; });
        request_1.request
            .post({
            url: 'https://www.google-analytics.com/collect',
            form: {
                v: 1,
                tid: 'UA-117778936-1',
                cid: Math.round(2147483647 * Math.random()).toString(),
                t: 'screenview',
                an: 'pact-install',
                av: require('../../../package.json').version,
                aid: 'pact-node',
                aiid: "standalone-" + versions_1.PACT_STANDALONE_VERSION,
                cd: "download-node-" + platform + "-" + (isCI ? 'ci' : 'user'),
                aip: true,
            },
        })
            .on('error', function () {
        });
    }
};
exports.trackDownload = trackDownload;
